src/aggregator/client/queue.go

```diff
@@ -204,22 +209,61 @@ func (q *queue) Close() error {
 	return nil
 }
 
+func (q *queue) writeAndReset() {
+	if len(q.buf) == 0 {
+		return
+	}
+	if err := q.writeFn(q.buf); err != nil {
+		q.log.Error("error writing data on timed flush",
+			zap.String("instance", q.instance.Endpoint()),
+			zap.Error(err),
+		)
+		q.metrics.connWriteErrors.Inc(1)
+	} else {
+		q.metrics.connWriteSuccesses.Inc(1)
+	}
+	q.buf = q.buf[:0]
+}
+
 func (q *queue) drain() {
 	defer q.wg.Done()
+	defer q.conn.Close()
+	timer := time.NewTimer(q.batchFlushDeadline)
+	lastDrain := time.Now()
```

Comment:
This is a super long lived function (background goroutine, right?)

In that case it seems reasonable to allocate a small helper function that you can use over and over:

```golang
write := func() {
    q.writeAndReset()
    lastDrain = time.Now()
}
```